nuget.exe should find the config file path based on the original working directory

## Details about Problem
Updating to a newer version on NuGet.exe breaks building solution. Errors show that it is looking for nuget.config in project folders when there already is a nuget.config at the solution level. 

Targets _IsProjectRestoreSupported and _GenerateRestoreProjectPathWalk seem to be needed. 

Command: NuGet.exe restore -ConfigFile src\.nuget\nuget.config -PackagesDirectory packages\ src\Project.sln

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): NuGet.exe and VS UI

NuGet version (x.x.x.xxx): 4.5.0.4487

VS version (if appropriate): 15.4.0 Preview 3.0 [26919.1.d15rel]

OS version (i.e. win10 v1607 (14393.321)): win10 v15063 (10.0.15063) 

Worked before? If so, with which NuGet version: 3.4.4.1321

## Other suggested things

### Verbose Logs
error MSB4057: The target "_GenerateRestoreProjectPathWalk" does not exist in the project.
error MSB4057: The target "_IsProjectRestoreSupported" does not exist in the project.
AppData\Local\Temp\NuGetScratch\bpguxuqr.e1t.nugetrestore.targets(476,5): error : File \InnerProject1\src\.nuget\nuget.config' does not exist.

--- a/src/NuGet.Core/NuGet.Build.Tasks/GetRestoreSettingsTask.cs
+++ b/src/NuGet.Core/NuGet.Build.Tasks/GetRestoreSettingsTask.cs
@@ -89,6 +89,7 @@ namespace NuGet.Build.Tasks
             BuildTasksUtility.LogInputParam(log, nameof(RestorePackagesPathOverride), RestorePackagesPathOverride);
             BuildTasksUtility.LogInputParam(log, nameof(RestoreSourcesOverride), RestoreSourcesOverride);
             BuildTasksUtility.LogInputParam(log, nameof(RestoreFallbackFoldersOverride), RestoreFallbackFoldersOverride);
+            BuildTasksUtility.LogInputParam(log, nameof(MSBuildStartupDirectory), MSBuildStartupDirectory);
 
             try
             {
@@ -108,18 +109,21 @@ namespace NuGet.Build.Tasks
                 }
 
                 // Settings
-                var settings = RestoreSettingsUtils.ReadSettings(RestoreSolutionDirectory, Path.GetDirectoryName(ProjectUniqueName), RestoreConfigFile, _machineWideSettings);
+                // Find the absolute path of nuget.config, this should only be set on the command line. Setting the path in project files
+                // is something that could happen, but it is not supported.
+                var absoluteConfigFilePath = GetGlobalAbsolutePath(RestoreConfigFile);
+                var settings = RestoreSettingsUtils.ReadSettings(RestoreSolutionDirectory, Path.GetDirectoryName(ProjectUniqueName), absoluteConfigFilePath, _machineWideSettings);
                 OutputConfigFilePaths = SettingsUtility.GetConfigFilePaths(settings).ToArray();
 
                 // PackagesPath
                 OutputPackagesPath = RestoreSettingsUtils.GetValue(
-                    () => string.IsNullOrEmpty(RestorePackagesPathOverride) ? null : UriUtility.GetAbsolutePath(MSBuildStartupDirectory, RestorePackagesPathOverride),
+                    () => GetGlobalAbsolutePath(RestorePackagesPathOverride),
                     () => string.IsNullOrEmpty(RestorePackagesPath) ? null : UriUtility.GetAbsolutePathFromFile(ProjectUniqueName, RestorePackagesPath),
                     () => SettingsUtility.GetGlobalPackagesFolder(settings));
 
                 // Sources
                 var currentSources = RestoreSettingsUtils.GetValue(
-                    () => RestoreSourcesOverride?.Select(MSBuildRestoreUtility.FixSourcePath).Select(e => UriUtility.GetAbsolutePath(MSBuildStartupDirectory, e)).ToArray(),
+                    () => RestoreSourcesOverride?.Select(MSBuildRestoreUtility.FixSourcePath).Select(e => GetGlobalAbsolutePath(e)).ToArray(),
                     () => MSBuildRestoreUtility.ContainsClearKeyword(RestoreSources) ? new string[0] : null,
                     () => RestoreSources?.Select(MSBuildRestoreUtility.FixSourcePath).Select(e => UriUtility.GetAbsolutePathFromFile(ProjectUniqueName, e)).ToArray(),
                     () => (new PackageSourceProvider(settings)).LoadPackageSources().Where(e => e.IsEnabled).Select(e => e.Source).ToArray());
@@ -136,7 +140,7 @@ namespace NuGet.Build.Tasks
 
                 // Fallback folders
                 var currentFallbackFolders = RestoreSettingsUtils.GetValue(
-                    () => RestoreFallbackFoldersOverride?.Select(e => UriUtility.GetAbsolutePath(MSBuildStartupDirectory, e)).ToArray(),
+                    () => RestoreFallbackFoldersOverride?.Select(e => GetGlobalAbsolutePath(e)).ToArray(),
                     () => MSBuildRestoreUtility.ContainsClearKeyword(RestoreFallbackFolders) ? new string[0] : null,
                     () => RestoreFallbackFolders?.Select(e => UriUtility.GetAbsolutePathFromFile(ProjectUniqueName, e)).ToArray(),
                     () => SettingsUtility.GetFallbackPackageFolders(settings).ToArray());
@@ -190,6 +194,19 @@ namespace NuGet.Build.Tasks
             }
 
             return items.SelectMany(e => MSBuildStringUtility.Split(BuildTasksUtility.GetPropertyIfExists(e, key)));
+        }
+
+        /// <summary>
+        /// Resolve a path against MSBuildStartupDirectory
+        /// </summary>
+        private string GetGlobalAbsolutePath(string path)
+        {
+            if (!string.IsNullOrEmpty(path))
+            {
+                return UriUtility.GetAbsolutePath(MSBuildStartupDirectory, path);
+            }
+
+            return path;
         }
     }
 }

--- a/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
+++ b/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
@@ -878,7 +878,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     ============================================================
   -->
   <Target Name="_GetRestoreSettingsOverrides"
-          Returns="$(_RestorePackagesPathOverride);$(_GetRestoreSourcesOverride);$(_GetRestoreFallbackFoldersOverride)">
+          Returns="$(_RestorePackagesPathOverride);$(_RestoreSourcesOverride);$(_RestoreFallbackFoldersOverride)">
 
     <!-- RestorePackagesPathOverride -->
     <MsBuild

--- a/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
+++ b/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
@@ -5573,5 +5573,59 @@ namespace NuGet.CommandLine.Test
 
             }
         }
+
+        [Fact]
+        public async Task RestoreNetCore_VerifyConfigFileWithRelativePathIsUsed()
+        {
+            // Arrange
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                // Set up solution, project, and packages
+                var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+
+                var projectA = SimpleTestProjectContext.CreateNETCore(
+                    "a",
+                    pathContext.SolutionRoot,
+                    NuGetFramework.Parse("net45"));
+
+                var packageX = new SimpleTestPackageContext()
+                {
+                    Id = "x",
+                    Version = "1.0.0"
+                };
+
+                projectA.AddPackageToAllFrameworks(packageX);
+
+                solution.Projects.Add(projectA);
+                solution.Create(pathContext.SolutionRoot);
+
+                var xml = File.ReadAllText(projectA.ProjectPath);
+                xml = xml.Replace("<TargetFrameworks>", "<TargetFramework>");
+                xml = xml.Replace("</TargetFrameworks>", "</TargetFramework>");
+                File.WriteAllText(projectA.ProjectPath, xml);
+
+                await SimpleTestPackageUtility.CreateFolderFeedV3(
+                    pathContext.PackageSource,
+                    PackageSaveMode.Defaultv3,
+                    packageX);
+
+                var subDir = Path.Combine(pathContext.SolutionRoot, "sub");
+                var configPath = Path.Combine(subDir, "nuget.config");
+                Directory.CreateDirectory(subDir);
+                File.Move(pathContext.NuGetConfig, configPath);
+
+                var relativePathToConfig = PathUtility.GetRelativePath(pathContext.WorkingDirectory + Path.DirectorySeparatorChar, configPath);
+
+                // Act
+                var r = Util.RestoreSolution(pathContext, 0, $"-ConfigFile {relativePathToConfig}");
+
+                // Assert
+                Assert.True(File.Exists(projectA.AssetsFileOutputPath), r.Item2);
+                Assert.True(File.Exists(projectA.TargetsOutput), r.Item2);
+                Assert.True(File.Exists(projectA.PropsOutput), r.Item2);
+
+                Assert.Equal(NuGetFramework.Parse("net45"), projectA.AssetsFile.Targets.Single(e => string.IsNullOrEmpty(e.RuntimeIdentifier)).TargetFramework);
+            }
+        }
     }
 }

